File: 1753-path-with-minimum-effort/path-with-minimum-effort.cpp
class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) 
    {
        int n = heights.size();
        int m = heights[0].size();
        
        // Directions for up, down, left, right movement
        vector<pair<int, int>> directions = {{0, -1}, {1, 0}, {0, 1}, {-1, 0}};
    
        // Distance vector to store minimum effort for each cell
        vector<vector<int>> effort(n, vector<int>(m, INT_MAX));
        effort[0][0] = 0; // Starting point distance is 0
    
        // Priority queue to process the smallest effort first
        // {effort, row, col}
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        pq.push({0, 0, 0}); 
    
        // BFS traversal
        while (!pq.empty()) 
        {
            auto curr = pq.top();
            pq.pop();
            int currEffort = curr[0];
            int row = curr[1];
            int col = curr[2];
        
            // If we reach the destination, return the distance
            if (row == n - 1 && col == m - 1)
            return effort[row][col];
        
            // Explore all possible directions
            for (auto [dx, dy] : directions) 
            {
                int newRow = row + dx;
                int newCol = col + dy;

                // Check if the new cell is within bounds
                if (newRow >= 0 && newRow < n && newCol >= 0 && newCol < m) 
                {
                    // Calculate the effort to move to the new cell
                    int newEffort = max(currEffort, abs(heights[row][col] - heights[newRow][newCol]));
                    // If a smaller effort is found, update and push to the priority queue
                    if (newEffort < effort[newRow][newCol]) 
                    {
                        effort[newRow][newCol] = newEffort;
                        pq.push({newEffort, newRow, newCol});
                    }
                }
            }
        }
        // return whatever the fuck you want to
        return -1;
    }
};

File: 1753-path-with-minimum-effort/path-with-minimum-effort.java
class Solution {
    private static int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    public int minimumEffortPath(int[][] grid) 
    {
        int rows = grid.length, cols = grid[0].length;
        boolean[][] visited = new boolean[rows][cols];

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.offer(new int[]{0, 0, 0});

        while (!pq.isEmpty()) 
        {
            int[] node = pq.poll();
            int r = node[1], c = node[2], currEffort = node[0];

            if(r == rows - 1 && c == cols - 1)
            return currEffort;

            if (visited[r][c]) continue;
            visited[r][c] = true;

            for (int[] dir : directions) 
            {
                int newRow = r + dir[0], newCol = c + dir[1];
                if(newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && !visited[newRow][newCol]) 
                {
                    int newEffort = Math.max(currEffort, Math.abs(grid[newRow][newCol] - grid[r][c]));
                    pq.offer(new int[]{newEffort, newRow, newCol});
                }
            }
        }
        return 0;
    }
}